In [1]:
from snowshovel.io import read_single_alert, parse_alerts
import pandas as pd
from tqdm import tqdm
from pathlib import Path

%matplotlib inline

Set up the config here for running the notebook

In [2]:
# Set this to the directory containing the unzipped avro alerts
data_dir = Path("/Users/robertstein/Data/ztf_public_20230609/")

# Set the number of alerts to read
n_to_read = 100

## What is an avro alert?

We can firstly look at a single alert, to see what it looks like:

In [3]:
alert_path = data_dir.joinpath("2350249170015010010.avro")
print(f"We will look at the avro alert with path '{alert_path}'")

We will look at the avro alert with path '/Users/robertstein/Data/ztf_public_20230609/2350249170015010010.avro'


In [4]:
alert = read_single_alert(alert_path)
print("An avro alert is just a python dictionary, as you can see below:")
alert

An avro alert is just a python dictionary, as you can see below:


{'jd': 2460104.7491782,
 'fid': 1,
 'pid': 2350249170015,
 'diffmaglim': 20.196775436401367,
 'pdiffimfilename': 'ztf_20230609249155_000326_zg_c01_o_q1_scimrefdiffimg.fits',
 'programpi': 'Kulkarni',
 'programid': 1,
 'candid': 2350249170015010010,
 'isdiffpos': 'f',
 'tblid': 10,
 'nid': 2350,
 'rcid': 0,
 'field': 326,
 'xpos': 1293.8194580078125,
 'ypos': 841.447509765625,
 'ra': 230.9148832,
 'dec': -19.2385929,
 'magpsf': 17.376073837280273,
 'sigmapsf': 0.047354377806186676,
 'chipsf': 14.490418434143066,
 'magap': 17.40329933166504,
 'sigmagap': 0.03269999846816063,
 'distnr': 0.16106265783309937,
 'magnr': 15.559000015258789,
 'sigmagnr': 0.013000000268220901,
 'chinr': 0.6050000190734863,
 'sharpnr': -0.017999999225139618,
 'sky': 0.5778443217277527,
 'magdiff': 0.027225999161601067,
 'fwhm': 2.9700000286102295,
 'classtar': 0.9950000047683716,
 'mindtoedge': 841.447509765625,
 'magfromlim': 2.793475389480591,
 'seeratio': 0.9369040727615356,
 'aimage': 0.8059999942779541,
 'b

## Handling many avro alerts:

That was just one avro alert, but our data will contain many sources, each with their own alert. python dictionaries are not very convenient for a large number of entries. Instead, we can use a function to read in a number of the alerts and convert them to a pandas dataframe.

In [5]:
alerts_df = parse_alerts(data_dir=data_dir, n_to_read=n_to_read)
print("Here you can see a pandas dataframe. The avro alert above is the first row in the table.")
alerts_df

100%|████████████████████████████████████████| 100/100 [00:00<00:00, 210.50it/s]

Here you can see a pandas dataframe. The avro alert above is the first row in the table.


jd  fid            pid  diffmaglim  \
0   2.460105e+06    1  2350249170015   20.196775   
1   2.460105e+06    1  2350249170015   20.196775   
2   2.460105e+06    1  2350249170015   20.196775   
3   2.460105e+06    1  2350249170015   20.196775   
4   2.460105e+06    1  2350249170015   20.196775   
..           ...  ...            ...         ...   
95  2.460105e+06    1  2350249170515   20.181540   
96  2.460105e+06    1  2350249170515   20.181540   
97  2.460105e+06    1  2350249170515   20.181540   
98  2.460105e+06    1  2350249170515   20.181540   
99  2.460105e+06    1  2350249170515   20.181540   

                                      pdiffimfilename programpi  programid  \
0   ztf_20230609249155_000326_zg_c01_o_q1_scimrefd...  Kulkarni          1   
1   ztf_20230609249155_000326_zg_c01_o_q1_scimrefd...  Kulkarni          1   
2   ztf_20230609249155_000326_zg_c01_o_q1_scimrefd...  Kulkarni          1   
3   ztf_20230609249155_000326_zg_c01_o_q1_scimrefd...  Kulkarni          1   
4   ztf_20230609249155_000326_zg_c01_o_q1_scimrefd...  Kulkarni          1   
..                                                ...       ...        ...   
95  ztf_20230609249155_000326_zg_c02_o_q2_scimrefd...  Kulkarni          1   
96  ztf_20230609249155_000326_zg_c02_o_q2_scimrefd...  Kulkarni          1   
97  ztf_20230609249155_000326_zg_c02_o_q2_scimrefd...  Kulkarni          1   
98  ztf_20230609249155_000326_zg_c02_o_q2_scimrefd...  Kulkarni          1   
99  ztf_20230609249155_000326_zg_c02_o_q2_scimrefd...  Kulkarni          1   

                 candid isdiffpos  tblid  ...      zpmed  clrmed    clrrms  \
0   2350249170015010010         f     10  ...  26.086000   0.554  0.230397   
1   2350249170015010011         f     11  ...  26.086000   0.554  0.230397   
2   2350249170015010013         f     13  ...  26.086000   0.554  0.230397   
3   2350249170015010023         f     23  ...  26.086000   0.554  0.230397   
4   2350249170015010024         f     24  ...  26.086000   0.554  0.230397   
..                  ...       ...    ...  ...        ...     ...       ...   
95  2350249170515015004         t      4  ...  26.224501   0.538  0.233582   
96  2350249170515015005         t      5  ...  26.224501   0.538  0.233582   
97  2350249170515015010         t     10  ...  26.224501   0.538  0.233582   
98  2350249170515015013         t     13  ...  26.224501   0.538  0.233582   
99  2350249170515015014         t     14  ...  26.224501   0.538  0.233582   

     neargaia  neargaiabright    maggaia  maggaiabright  exptime       drb  \
0    0.228274     -999.000000  15.315500    -999.000000     30.0  0.999971   
1    0.066207     -999.000000  14.556261    -999.000000     30.0  0.999255   
2    0.175491     -999.000000  16.767410    -999.000000     30.0  1.000000   
3    0.398773     -999.000000  16.487196    -999.000000     30.0  1.000000   
4    0.105536       52.636383  18.771164      13.606915     30.0  0.999883   
..        ...             ...        ...            ...      ...       ...   
95  15.838425       76.763954  18.365631      13.423489     30.0  0.999845   
96  16.460882     -999.000000  18.606089    -999.000000     30.0  1.000000   
97  14.950565     -999.000000  19.616867    -999.000000     30.0  0.999972   
98   5.632905       80.984413  20.652666      13.806736     30.0  0.999986   
99   0.314002       46.619858  18.730354      13.724556     30.0  0.999894   

    drbversion  
0        d6_m7  
1        d6_m7  
2        d6_m7  
3        d6_m7  
4        d6_m7  
..         ...  
95       d6_m7  
96       d6_m7  
97       d6_m7  
98       d6_m7  
99       d6_m7  

[100 rows x 103 columns]

The table above contains many rows and columns. It can be hard to see what the important information. Let's just display a few relevant columns, rather than all 103.

In [6]:
useful_columns = [
    "isdiffpos", # Is the detection postive (more flux in science image than reference), or negative (less flux)?
    "ra", # What is the sky position in Right Ascension?
    "dec", # What is the sky position in Declination?
    "magpsf", # What is the magnitude (brightness) of the detection?
    "sigmapsf", # What is the uncertainty on the magnitude?
    "distnr", # How close is the detection to a known source in the reference image?
    "ndethist", # How many times has this object been detected?
]
slim_alerts_df = alerts_df[useful_columns]
slim_alerts_df

isdiffpos          ra        dec     magpsf  sigmapsf     distnr  ndethist
0          f  230.914883 -19.238593  17.376074  0.047354   0.161063       620
1          f  230.630090 -19.273569  17.239075  0.078380   0.173305       639
2          f  231.014147 -19.282204  18.312075  0.061234   0.134472       611
3          f  230.909926 -19.485533  18.983074  0.085728   0.424015       721
4          f  230.435069 -19.499763  20.108074  0.164924   0.145266        65
..       ...         ...        ...        ...       ...        ...       ...
95         t  228.200989 -19.285597  19.772402  0.170230  15.738925         1
96         t  227.781890 -19.395761  19.201401  0.113029  12.420472         1
97         t  228.044432 -19.503497  19.268400  0.112572   4.766288         1
98         t  228.354661 -19.831578  18.597401  0.085115   4.054725         1
99         t  228.226035 -19.828269  19.959400  0.212788   0.239048       191

[100 rows x 7 columns]

In [9]:
# Bad subtractions are often not repeated
# Let's restrict ourselves to sources which are detected at least twice
mask = slim_alerts_df["ndethist"] > 1
slim_alerts_df = slim_alerts_df[mask].reset_index()
print(f"There are {len(slim_alerts_df)} dources ")

slim_alerts_df

index isdiffpos          ra        dec     magpsf  sigmapsf    distnr  \
0       0         f  230.914883 -19.238593  17.376074  0.047354  0.161063   
1       1         f  230.630090 -19.273569  17.239075  0.078380  0.173305   
2       2         f  231.014147 -19.282204  18.312075  0.061234  0.134472   
3       3         f  230.909926 -19.485533  18.983074  0.085728  0.424015   
4       4         f  230.435069 -19.499763  20.108074  0.164924  0.145266   
5       7         t  230.453158 -19.123650  18.570074  0.079196  0.143341   
6      11         t  231.242650 -19.245702  16.856073  0.091851  0.708412   
7      17         f  230.045208 -19.158488  17.748367  0.086055  0.774607   
8      18         f  229.473594 -19.301989  18.538366  0.186850  0.283202   
9      19         f  230.072448 -19.420921  19.711367  0.144221  0.132384   
10     20         f  229.807137 -19.792181  20.061367  0.201591  0.356555   
11     21         f  229.596090 -19.645596  18.389366  0.088047  0.168681   
12     27         t  229.722340 -19.784506  18.697367  0.118325  0.325324   
13     31         f  230.150072 -20.062295  20.141975  0.201107  1.121228   
14     32         f  230.181551 -20.113803  19.993975  0.191287  2.946234   
15     33         f  230.315938 -20.171301  19.121975  0.172073  0.388920   
16     34         f  230.271784 -20.592282  16.587975  0.114430  0.488886   
17     36         t  229.687871 -19.920162  19.946976  0.186163  0.330008   
18     38         t  230.047006 -19.952832  19.609976  0.147735  0.466755   
19     40         t  229.739482 -20.014014  17.667974  0.053372  9.750773   
20     49         t  229.923247 -20.602498  19.868975  0.196410  0.563099   
21     50         t  230.111051 -20.452607  16.320974  0.050384  0.074213   
22     51         f  230.474714 -19.936488  19.463905  0.124944  0.257316   
23     52         f  231.175496 -20.121978  19.568905  0.216569  0.630820   
24     53         f  230.621346 -20.266582  17.015905  0.039268  0.213959   
25     54         f  230.952607 -20.343225  15.621905  0.055531  0.252521   
26     55         f  230.997230 -20.484602  17.924906  0.053547  0.076702   
27     58         t  231.199842 -20.100425  19.617905  0.137637  0.208387   
28     59         t  231.010622 -20.115624  15.068905  0.023402  0.116662   
29     61         t  230.709022 -20.278496  19.401905  0.141603  0.305033   
30     62         t  230.727811 -20.338037  18.603905  0.122564  0.770484   
31     65         t  230.669811 -20.463484  15.622905  0.028162  1.232800   
32     67         f  228.799087 -19.058811  18.699652  0.104450  0.530823   
33     68         f  228.531597 -19.196783  19.214651  0.107263  0.293129   
34     69         f  229.140051 -19.883046  20.121653  0.212115  0.102527   
35     70         f  228.903759 -19.835699  17.597652  0.112485  0.620152   
36     73         t  228.704875 -19.141045  19.351652  0.142213  0.289885   
37     77         t  228.610251 -19.299282  16.588652  0.050217  0.456356   
38     78         t  229.249739 -19.436477  17.907652  0.059456  0.105491   
39     80         t  228.636061 -19.477837  19.154652  0.115297  0.621792   
40     81         t  229.160634 -19.864842  16.760653  0.046199  0.339500   
41     85         t  228.846211 -19.666244  18.091652  0.059456  0.170502   
42     86         t  229.194059 -19.603891  18.961653  0.151453  2.288612   
43     88         t  229.084361 -19.609338  19.968653  0.161095  0.258935   
44     89         f  228.270052 -19.110670  20.341400  0.211415  0.231607   
45     90         f  227.638027 -19.212167  15.857401  0.026999  0.234082   
46     91         f  228.250654 -19.264291  19.378401  0.128130  0.156076   
47     92         f  227.800896 -19.773364  18.773401  0.199060  0.268053   
48     94         t  227.988595 -19.225825  17.047401  0.055371  0.073341   
49     99         t  228.226035 -19.828269  19.959400  0.212788  0.239048   

    ndethist  
0        620  
1        639  
2        611  
3        721  
4